APMrover2/sailboat.cpp

```diff
@@ -20,138 +20,213 @@ To Do List
  - add option to do proper tacks, ie tacking on flat spot in the waves, or only try once at a certain speed, or some better method than just changing the desired heading suddenly
 */
 
+const AP_Param::GroupInfo Sailboat::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Enable Sailboat
+    // @Description: This enables Sailboat functionality
+    // @Values: 0:Disable,1:Enable
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO_FLAGS("ENABLE", 1, Sailboat, enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: ANGLE_MIN
+    // @DisplayName: Sail min angle
+    // @Description: Mainsheet tight, angle between centerline and boom
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("ANGLE_MIN", 2, Sailboat, sail_angle_min, 0),
+
+    // @Param: ANGLE_MAX
+    // @DisplayName: Sail max angle
+    // @Description: Mainsheet loose, angle between centerline and boom
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("ANGLE_MAX", 3, Sailboat, sail_angle_max, 90),
+
+    // @Param: ANGLE_IDEAL
+    // @DisplayName: Sail ideal angle
+    // @Description: Ideal angle between sail and apparent wind
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("ANGLE_IDEAL", 4, Sailboat, sail_angle_ideal, 25),
+
+    // @Param: HEEL_MAX
+    // @DisplayName: Sailing maximum heel angle
+    // @Description: When in auto sail trim modes the heel will be limited to this value using PID control
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("HEEL_MAX", 5, Sailboat, sail_heel_angle_max, 15),
+
+    // @Param: SAIL_NO_GO_ANGLE
+    // @DisplayName: Sailing no go zone angle
+    // @Description: The typical closest angle to the wind the vehicle will sail at. the vehicle will sail at this angle when going upwind
+    // @Units: deg
+    // @Range: 0 90
+    // @Increment: 1
+    // @User: Standard
+    AP_GROUPINFO("NO_GO_ANGLE", 6, Sailboat, sail_no_go, 45),
+
+    AP_GROUPEND
+};
+
+
+/*
+  constructor
+ */
+Sailboat::Sailboat()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+
+    // sailboat defaults
+    if (enabled()) {
```

Comment:
Have you tested that the defaults are taking effect? At this point the parameter values shouldn't be loaded so enabled should always return false. You likely need an init method, where you should also do the check that `has_sail` did - if the output isn't configured then we should set (not save) enable to 0.

Reply:
I did test it and it worked OK, the init method seems like a better way to go I will do as you suggest, thanks

Reply:
done